salt/_modules/metalk8s_checks.py

```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+            'mandatory': 'True/False'
+        }
+    }
+    ```
+    Where:
+    - `<address>` could be a full address `<ip>:<port>` or just a `<port>`
+      (if `<ip>` is equal to `control_plane_ip` or `workload_plane_ip` we replace
+      it with the local expected IP
+    - `<processes>` could be a single process regexp matching or a list of regexp
+      (if one of the processes in this list match then result is ok)
+    - `<description>` is optional, and is just a description of the expected usage
+      of this address
+    - `<mandatory>` is a boolean to force this expected process, if set to False
+      we expected either, the process in `expected` either nothing (default: False)
+    """
+    if listening_process is None:
+        if listening_process_per_role is None:
+            listening_process_per_role = __salt__["metalk8s.get_from_map"](
+                "networks", saltenv=kwargs.get("saltenv")
+            )["listening_process_per_role"]
+        if roles is None:
+            roles = (
+                __pillar__.get("metalk8s", {})
+                .get("nodes", {})
+                .get(__grains__["id"], {})
+                .get("roles")
+            )
+
+        # If role not yet set consider we have all roles
+        if roles is None:
+            roles = listening_process_per_role.keys()
+
+        # Compute full dict of listening process according to local `roles`
+        # Note: We consider all minions as "node"
+        listening_process = listening_process_per_role.get("node") or {}
+        for role in roles:
+            listening_process.update(listening_process_per_role.get(role) or {})
+
+    if not isinstance(listening_process, dict):
+        raise ValueError(
+            "Invalid listening process, expected dict but got {}.".format(
+                type(listening_process).__name__
+            )
+        )
+
+    errors = []
+
+    all_listen_connections = __salt__["metalk8s_network.get_listening_processes"]()
+
+    for address, matching in listening_process.items():
+        if isinstance(address, int):
+            address = str(address)
+        ip = None
+        port = address
+        if len(address.rsplit(":")) == 2:
+            ip, port = address.rsplit(":")
+
+        if ip and ip in ["control_plane_ip", "workload_plane_ip"]:
+            ip = __grains__["metalk8s"][ip]
+
+            # We also update the `address` for error message
+            address = "{}:{}".format(ip, port)
+
+        processes = matching.get("expected")
+        if not isinstance(processes, list):
+            processes = [processes]
+
+        if not matching.get("mandatory") and None not in processes:
+            processes.append(None)
+
+        process_on_port = all_listen_connections.get(
+            str(port), all_listen_connections.get(int(port))
+        )
+
+        success = False
+        error_process = {}
+        # If one of this process is expected one, succeed
+        for process in processes:
```

Comment:
Yes exactly but note also that if you would have put the None manually and got a 
```python
processes = [None, "kube-apiserver"]
```
The result would have been the same, just error for None if apiserver listening but this error is ignored in the resulting message anyway since one of the process match.

I will add a comment